In [ ]:
# Generadores
Una serie de problemas para ser resueltos con generadores. Seguramente necesitaras repasar un poco generadores, pero es parte del aprendizaje.  Son cinco pasos en total.  Cada paso es independiente del siguiente, pero puede resultar útil lo que has hecho en el paso anterior.
En cada bloque, debes definir la función para que el resto del código ya escrito funcione y genere el output esperado.
Copia el notebook en local, modifícalo (pruebalo muchas veces) y cuando lo tengas listo, súbelo al campus virtual o a tu propio github y comparte el enlace en la tarea.

Parte 1: Crea un generador que calcule la media corriente (running average)

In [1]:
def running_average():
    
    values = [0, 7, 13, 17, 231, 12, 8, 3]

    sum_is = 0
    for i in range(len(values)):   
        if i == 0:
            print("Division by 0")
            yield(0)
        else: 
            sum_is = sum_is + values[i]
            avg = sum_is/(i)
            
            yield(avg)
     ###   
    
ra = running_average() 
next(ra)                
for value in [7, 13, 17, 231, 12, 8, 3]:
    out_str = "sent: {val:3d}, new average: {avg:6.2f}"
    print(out_str.format(val=value, avg=ra.send(value)))  

División por cero a ocurrido
sent:   7, new average:   7.00
sent:  13, new average:  10.00
sent:  17, new average:  12.33
sent: 231, new average:  67.00
sent:  12, new average:  56.00
sent:   8, new average:  48.00
sent:   3, new average:  41.57


Esta es la salida esperada de tu programa<br>
sent:   7, new average:   7.00<br>
sent:  13, new average:  10.00<br>
sent:  17, new average:  12.33<br>
sent: 231, new average:  67.00<br>
sent:  12, new average:  56.00<br>
sent:   8, new average:  48.00<br>
sent:   3, new average:  41.57

Parte 2: escribe un generador "trange", que genere una secuencia de tuplas de tiempo desde un "inicio" hasta un "final" incrementados por un "paso". Una tupla de tiempo es una tupla de 3 enteros: (horas,minutos,segundos)

In [1]:
def trange(start, stop, step):
  
    
    

    tsum = start 
    lsum = start 
    
 
    subt    = [0,0,0]
    stopl    = list(stop)
    stepl    = list(step)
    subt[0] = stopl[0] - stepl[0]
    subt[1] = stopl[1] - stepl[1]
    subt[2] = stopl[2] - stepl[2]
    restat   = tuple(subt) 

    print(start) 
    while tsum<=subt: 
        stepl  = list(step)   
        lsum = [lsum, stepl]
        lsum = [sum(i) for i in zip(*lsum)]  
        if lsum[2] == 60: 
            lsum[1] = lsum[1] + 1
            lsum[2] = 0           
        elif lsum[2]>60: 
            lsum[2] = lsum[2] - 60
            lsum[1] = lsum[1] + 1
        if lsum[1] == 60:
            lsum[0] = lsum[0] + 1
            lsum[1] = 0      
        elif lsum[1] >60:
            lsum[1] = -60 + lsum[1]
            lsum[0] = lsum[0] + 1
        if lsum[0] >= 60: 
            lsum[0] == 0 
            lsum[1] == 0
            lsum[2] == 0

        tsum = tuple(lsum) 
        yield(tsum)
    
    
    
for time in trange((12,15,10),(14,10,10),(0,15,12)):
    print(time)
    

(12, 15, 10)
(12, 30, 22)
(12, 45, 34)
(13, 0, 46)
(13, 15, 58)
(13, 31, 10)
(13, 46, 22)
(14, 1, 34)


Salida esperada:<br>
(12, 15, 10)<br>
(12, 30, 22)<br>
(12, 45, 34)<br>
(13, 0, 46)<br>
(13, 15, 58)<br>
(13, 31, 10)<br>
(13, 46, 22)<br>
(14, 1, 34)

In [3]:
def rtrange(start, stop, step):
    
    
    tsum = start 
    lsum = start 
    

    subt    = [0,0,0]
    stopl    = list(stop)
    stepl    = list(step)
    subt[0] = stopl[0] - stepl[0]
    subt[1] = stopl[1] - stepl[1]
    subt[2] = stopl[2] - stepl[2]
    restat   = tuple(subt) 

    print(start) 
    while tsum<=subt: 
    
        stepl  = list(step)   
        lsum = [lsum, stepl]
        lsum = [sum(i) for i in zip(*lsum)]  
        if lsum[2] == 60:
            lsum[1] = lsum[1] + 1
            lsum[2] = 0           
        elif lsum[2]>60: 
            lsum[2] = lsum[2] - 60
            lsum[1] = lsum[1] + 1
        if lsum[1] == 60:
            lsum[0] = lsum[0] + 1
            lsum[1] = 0      
        elif lsum[1] >60:
            lsum[1] = -60 + lsum[1]
            lsum[0] = lsum[0] + 1
        if lsum[0] >= 23: 
            lsum[0] == 0 
            lsum[1] == 0
            lsum[2] == 0

        tsum = tuple(lsum) 
        
        
        new_tsum = yield(tsum)
        if(new_tsum != None):
            lsum = new_tsum
            print(lsum)
    
    ####
    
ts = rtrange((10, 10, 10), (13, 50, 15), (0, 15, 12) )  
for _ in range(3):
    print(next(ts))
        
print(ts.send((8, 5, 50)))
for _ in range(3):
    print(next(ts))
    

(10, 10, 10)
(10, 25, 22)
(10, 40, 34)
(10, 55, 46)
(8, 5, 50)
(8, 21, 2)
(8, 36, 14)
(8, 51, 26)
(9, 6, 38)


Salida esperada:<br>
(10, 10, 10)<br>
(10, 25, 22)<br>
(10, 40, 34)<br>
(8, 5, 50)<br>
(8, 21, 2)<br>
(8, 36, 14)<br>
(8, 51, 26)

Parte 4: utilice el programa trange para generar un archivo "Tiempo_vs_Temperatura.txt". Cada línea del archivo debe contener un tiempo en formato hh:mm:ss y una temperatura (aleatoria) entre 10 y 25 graods. Los tiempos deben ser en orden ascendente con un intervalo de 90 segundos, empezando desde 06:00:00 y terminando a las 23:00:00. No es necesario reescribir la función trange; solo compilarla antes para poderla usar aquí.

In [3]:

import random

def trange(start, stop, step):



    tsum = start 
    lsum = start 
    

    subt    = [0,0,0]
    stopl    = list(stop)
    stepl    = list(step)
    subt[0] = stopl[0] - stepl[0]
    subt[1] = stopl[1] - stepl[1]
    subt[2] = stopl[2] - stepl[2]
    restat   = tuple(subt) 

    while tsum<=restat: 
        stepl  = list(step)   
        lsum = [lsum, stepl]
        lsum = [sum(i) for i in zip(*lsum)]  
        if lsum[2] == 60: 
            lsum[1] = lsum[1] + 1
            lsum[2] = 0           
        elif lsum[2]>60: 
            lsum[2] = lsum[2] - 60
            lsum[1] = lsum[1] + 1
        if lsum[1] == 60:
            lsum[0] = lsum[0] + 1
            lsum[1] = 0      
        elif lsum[1] >60:
            lsum[1] = -60 + lsum[1]
            lsum[0] = lsum[0] + 1
        if lsum[0] >= 23: 
            lsum[0] == 0 
            lsum[1] == 0
            lsum[2] == 0

        tsum = tuple(lsum) 
        yield(tsum)
        
#se manda a imprimir los valores con los nuevos parámetros start, stop, step. 
for timer in trange((6,0,0),(22,58,30),(0, 1 ,30)):
    outStr = '{:02d}:{:02d}:{:02d} {} {}' 
    salida = (outStr.format(timer[0],timer[1],timer[2], "    ", round(random.uniform(10,25),1)))
    
    """
    Para crear el archivo.txt con los valores se manada el valor en print
    
    """    
    archivo = open("Tiempo_vs_Temperatura.txt", "a")
    archivo.write("\n")
    archivo.writelines(salida)
    archivo.close()






El archivo debe contener algo parecido a esto (la temperatura es aleatoria, pero los tiempos deben ser iguales a estos):

06:00:00 14.0

06:01:30 10.5

06:03:00 21.0

...

22:57:00 12.1

22:58:30 21.5

##end of file

In [29]:
import random

def random_ones_and_zeros():
    ### escribe aquí el codigo, recuerda que p=0.5
    p = 0.5
    while p!=0: 
        number = random.randint(0,100)

        if number < p*100:             
            bit = yield(1)
        else: 
            bit = yield(0)
    ###
    
x = random_ones_and_zeros()
next (x)    
for p in [0.2,0.5,0.8]:  
    x.send(p)
    print("\n probability: "+str(p))
    for i in range(15):
        print(next(x), end=" ")



 probability: 0.2
0 0 0 0 0 1 0 0 0 0 0 0 1 0 0 
 probability: 0.5
0 0 1 1 1 1 1 0 1 1 1 1 0 0 0 
 probability: 0.8
1 1 0 0 0 1 0 0 0 0 0 0 1 1 1 

La salida esperada debe ser similar a esta:<br>
probability: 0.2<br>
0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 <br>
probability: 0.5<br> 
1 0 0 1 0 0 0 1 0 1 1 0 0 0 1<br>
probability: 0.8<br>
1 1 1 1 1 1 1 1 1 1 0 0 1 1 1 
